cookbooks/anisotropic_viscosity/av_material.cc

```diff
@@ -0,0 +1,867 @@
+        virtual void initialize();
+        virtual void evaluate (const MaterialModel::MaterialModelInputs<dim> &in,
+                               MaterialModel::MaterialModelOutputs<dim> &out) const;
+        static void declare_parameters (ParameterHandler &prm);
+        virtual void parse_parameters (ParameterHandler &prm);
+        virtual bool is_compressible () const;
+        virtual double reference_viscosity () const;
+        virtual double reference_density () const;
+        virtual void create_additional_named_outputs(MaterialModel::MaterialModelOutputs<dim> &out) const;
+      private:
+        double eta_N, viscosity_ratio; //normal viscosity and ratio between the shear and the normal viscosities
+        static int delta (const unsigned int i, const unsigned int j); //kronecker delta function
+        void set_assemblers(const SimulatorAccess<dim> &,
+                            Assemblers::Manager<dim> &assemblers) const;
+    };
+  }
+}
+
+namespace aspect
+{
+
+//Next session is a more evolved implementation of anisotropic viscosity in the material model by Jonathan Perry-Houts
+  namespace MaterialModel
+  {
+    template <int dim>
+    void
+    AV<dim>::set_assemblers(const SimulatorAccess<dim> &,
+                            Assemblers::Manager<dim> &assemblers) const
+    {
+      for (unsigned int i=0; i<assemblers.stokes_preconditioner.size(); ++i)
+        {
+          if (dynamic_cast<Assemblers::StokesPreconditioner<dim> *>(assemblers.stokes_preconditioner[i].get()) != nullptr)
+            assemblers.stokes_preconditioner[i] = std_cxx14::make_unique<Assemblers::StokesPreconditionerAnisotropicViscosity<dim> > ();
+        }
+
+      for (unsigned int i=0; i<assemblers.stokes_system.size(); ++i)
+        {
+          if (dynamic_cast<Assemblers::StokesIncompressibleTerms<dim> *>(assemblers.stokes_system[i].get()) != nullptr)
+            assemblers.stokes_system[i] = std_cxx14::make_unique<Assemblers::StokesIncompressibleTermsAnisotropicViscosity<dim> > ();
+        }
+    }
+
+
+
+    template <int dim>
+    void
+    AV<dim>::
+    initialize()
+    {
+      this->get_signals().set_assemblers.connect (std::bind(&AV<dim>::set_assemblers,
+                                                            std::cref(*this),
+                                                            std::placeholders::_1,
+                                                            std::placeholders::_2));
+      AssertThrow((dim==2),
+                  ExcMessage("For now the anisotropic viscosity assemblers work only in 2D"));
+
+    }
+
+
+
+    template <int dim>
+    void
+    AV<dim>::evaluate (const MaterialModel::MaterialModelInputs<dim> &in,
+                       MaterialModel::MaterialModelOutputs<dim> &out) const
+    {
+      MaterialModel::AnisotropicViscosity<dim> *anisotropic_viscosity =
+        out.template get_additional_output<MaterialModel::AnisotropicViscosity<dim> >();
+
+      //Simple<dim>::evaluate(in,out);
```

Comment:
So now I'm confused. You derive the class from the `Simple` model, but you don't actually use the `Simple` implementation anywhere, do you? Why not call the function you have commented out here? You overwrite various variables below, but I'm not clear why you need to do that.

Alternatively, if you don't want to build on the `Simple` model, then remove this line, remove the derivation of this class from `Simple`, and update the comment in front of the class that mentions `Simple`.

You'll have to do the same for the other material model.